js_modules/dagit/packages/core/src/instance/InstanceConcurrency.tsx

```diff
@@ -0,0 +1,530 @@
+                      onDelete={onDelete}
+                    />
+                  </td>
+                </tr>
+              ))}
+            </tbody>
+          </Table>
+        )}
+      </Box>
+      <AddConcurrencyLimitDialog
+        open={!!action && action.actionType === 'add'}
+        onClose={() => setAction(undefined)}
+        onComplete={refetch}
+      />
+      <DeleteConcurrencyLimitDialog
+        concurrencyKey={action && action.actionType === 'delete' ? action.concurrencyKey : ''}
+        open={!!action && action.actionType === 'delete'}
+        onClose={() => setAction(undefined)}
+        onComplete={refetch}
+      />
+      <EditConcurrencyLimitDialog
+        open={!!action && action.actionType === 'edit'}
+        onClose={() => setAction(undefined)}
+        onComplete={refetch}
+        concurrencyKey={action && action.actionType === 'edit' ? action.concurrencyKey : ''}
+      />
+      <ConcurrencyRunsDialog
+        title={
+          <span>
+            Active runs for <strong>{selectedKey}</strong>
+          </span>
+        }
+        onClose={() => {
+          setSelectedRuns(undefined);
+          setSelectedKey(undefined);
+        }}
+        runIds={selectedRuns}
+      />
+    </>
+  );
+};
+
+const ConcurrencyLimitActionMenu: React.FC<{
+  concurrencyKey: string;
+  onEdit: (key: string) => void;
+  onDelete: (key: string) => void;
+}> = ({concurrencyKey, onDelete, onEdit}) => {
+  return (
+    <Popover
+      content={
+        <Menu>
+          <MenuItem icon="edit" text="Edit" onClick={() => onEdit(concurrencyKey)} />
+          <MenuItem
+            icon="delete"
+            intent="danger"
+            text="Delete"
+            onClick={() => onDelete(concurrencyKey)}
+          />
+        </Menu>
+      }
+      position="bottom-left"
+    >
+      <Button icon={<Icon name="expand_more" />} />
+    </Popover>
+  );
+};
+
+const isValidLimit = (concurrencyLimit?: string) => {
+  if (!concurrencyLimit) {
+    return false;
+  }
+  const value = parseInt(concurrencyLimit);
+  if (isNaN(value)) {
+    return false;
+  }
+  if (String(value) !== concurrencyLimit.trim()) {
+    return false;
+  }
+  return value > 0 && value < 1000;
+};
+
+const AddConcurrencyLimitDialog: React.FC<{
+  open: boolean;
+  onClose: () => void;
+  onComplete: () => void;
+}> = ({open, onClose, onComplete}) => {
+  const [isSubmitting, setIsSubmitting] = React.useState(false);
+  const [limitInput, setLimitInput] = React.useState('');
+  const [keyInput, setKeyInput] = React.useState('');
+
+  React.useEffect(() => {
+    setLimitInput('');
+    setKeyInput('');
+  }, [open]);
+
+  const [setConcurrencyLimit] = useMutation<
+    SetConcurrencyLimitMutation,
+    SetConcurrencyLimitMutationVariables
+  >(SET_CONCURRENCY_LIMIT_MUTATION);
+
+  const save = async () => {
+    setIsSubmitting(true);
+    await setConcurrencyLimit({
+      variables: {concurrencyKey: keyInput, limit: parseInt(limitInput!.trim())},
```

Comment:
I don't think you should need the `!` on `limitInput` since it's always a string